<a href="https://colab.research.google.com/github/640710539pawarisa/ai_lab_/blob/main/lab_week5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ตอนที่ 8สมองจิ๋วเรียน XOR ต่อหน้าต่อตา


In [41]:
import numpy as np

rng = np.random.default_rng(1)
X = np.array([[0,0],[0,1],[1,0],[1,1]], float)
y = np.array([[0],[1],[1],[0]], float)

W1 = rng.normal(0, 1, (2,2)); b1 = np.zeros((1,2))
W2 = rng.normal(0, 1, (2,1)); b2 = np.zeros((1,1))
sig = lambda z: 1/(1+np.exp(-z))

lr = 0.5
for epoch in range(20001):
    h = sig(X@W1 + b1)          # ชั้นซ่อนคิด
    out = sig(h@W2 + b2)        # ชั้นตอบคิด
    loss = np.mean((out-y)**2)  # วัดความผิด

    d_out = (out-y)*out*(1-out)    # ใบตำหนิชั้นตอบ
    d_h = d_out@W2.T * h*(1-h)     # ส่งย้อนไปชั้นซ่อน
    W2 -= lr*h.T@d_out; b2 -= lr*d_out.sum(0)
    W1 -= lr*X.T@d_h;  b1 -= lr*d_h.sum(0)

In [42]:
x = 8.0
lr = 0.1                      # ขนาดก้าว

for i in range(21):
    if i in (0, 1, 2, 3, 5, 10, 20):
        print(f"ก้าว {i:2d}  x = {x:7.4f}  ความสูง = {x*x:8.4f}")
    gradient = 2*x            # ความชันใต้ฝ่าเท้า
    x = x - lr*gradient       # ก้าวสวนทางความชัน

ก้าว  0  x =  8.0000  ความสูง =  64.0000
ก้าว  1  x =  6.4000  ความสูง =  40.9600
ก้าว  2  x =  5.1200  ความสูง =  26.2144
ก้าว  3  x =  4.0960  ความสูง =  16.7772
ก้าว  5  x =  2.6214  ความสูง =   6.8719
ก้าว 10  x =  0.8590  ความสูง =   0.7379
ก้าว 20  x =  0.0922  ความสูง =   0.0085


In [43]:
for lr in [0.01, 0.1, 0.45, 0.55, 1.05]:
    x = 8.0
    for _ in range(20):
        x = x - lr*2*x
    print(f"lr {lr:5.2f}  หลัง 20 ก้าว x = {x:.4f}")

lr  0.01  หลัง 20 ก้าว x = 5.3409
lr  0.10  หลัง 20 ก้าว x = 0.0922
lr  0.45  หลัง 20 ก้าว x = 0.0000
lr  0.55  หลัง 20 ก้าว x = 0.0000
lr  1.05  หลัง 20 ก้าว x = 53.8200


In [44]:
import numpy as np

rng = np.random.default_rng(1)
X = np.array([[0,0],[0,1],[1,0],[1,1]], float)
y = np.array([[0],[1],[1],[0]], float)



In [45]:
y

array([[0.],
       [1.],
       [1.],
       [0.]])

In [46]:
W1 = rng.normal(0, 1, (2,2)); b1 = np.zeros((1,2))
W2 = rng.normal(0, 1, (2,1)); b2 = np.zeros((1,1))
sig = lambda z: 1/(1+np.exp(-z))


In [47]:
W1

array([[ 0.34558419,  0.82161814],
       [ 0.33043708, -1.30315723]])

In [48]:
W2

array([[0.90535587],
       [0.44637457]])

In [49]:

h = sig(X@W1 + b1)            # ชั้นซ่อนคิด
out = sig(h@W2 + b2)          # ชั้นตอบคิด


In [50]:
print("คำตอบก่อนฝึก", np.round(out.ravel(), 3))
print("loss ก่อนฝึก", round(float(np.mean((out-y)**2)), 4))

คำตอบก่อนฝึก [0.663 0.651 0.699 0.684]
loss ก่อนฝึก 0.2799


In [51]:
lr = 0.5
for epoch in range(20001):
    h = sig(X@W1 + b1)                  # จังหวะ 1 คิดไปข้างหน้า
    out = sig(h@W2 + b2)
    loss = np.mean((out-y)**2)          # จังหวะ 2 วัดความผิด
    if epoch in (0, 100, 1000, 5000, 20000):
        print(f"รอบ {epoch:5d}  loss = {loss:.4f}")


##ตรงนี้คือ dif แบบลูกโซ่
    d_out = (out-y)*out*(1-out)         # จังหวะ 3 ใบตำหนิชั้นตอบ
    d_h = d_out@W2.T * h*(1-h)          # ส่งย้อนไปชั้นซ่อน

    ##
    W2 -= lr*h.T@d_out; b2 -= lr*d_out.sum(0)
    W1 -= lr*X.T@d_h;  b1 -= lr*d_h.sum(0)

print("คำตอบหลังฝึก", np.round(out.ravel(), 3))

รอบ     0  loss = 0.2799
รอบ   100  loss = 0.2486
รอบ  1000  loss = 0.0155
รอบ  5000  loss = 0.0007
รอบ 20000  loss = 0.0001
คำตอบหลังฝึก [0.013 0.989 0.989 0.011]


In [52]:
print(np.round(sig(X@W1 + b1), 2))


[[0.04 0.03]
 [0.93 0.  ]
 [0.   0.95]
 [0.03 0.02]]


In [57]:
import torch
import torch.nn as nn

In [58]:
torch.manual_seed(1)
X = torch.tensor([[0.,0.],[0.,1.],[1.,0.],[1.,1.]])
y = torch.tensor([[0.],[1.],[1.],[0.]])

In [59]:

model = nn.Sequential(nn.Linear(2,2), nn.Sigmoid(),
                      nn.Linear(2,1), nn.Sigmoid())
opt = torch.optim.SGD(model.parameters(), lr=0.5)
loss_fn = nn.MSELoss()


In [60]:

for epoch in range(20001):
    out = model(X)                      # คิดไปข้างหน้า
    loss = loss_fn(out, y)              # วัดความผิด
    if epoch in (0, 100, 1000, 5000, 20000):
        print(f"รอบ {epoch:5d}  loss = {loss.item():.4f}")
    opt.zero_grad()
    loss.backward()                     # ใบตำหนิย้อนกลับ ในบรรทัดเดียว
    opt.step()                          # ขยับทุกน้ำหนักหนึ่งก้าว



รอบ     0  loss = 0.2548
รอบ   100  loss = 0.2501
รอบ  1000  loss = 0.2500
รอบ  5000  loss = 0.0355
รอบ 20000  loss = 0.0004


In [61]:
print("คำตอบหลังฝึก", model(X).detach().numpy().round(3).ravel())

คำตอบหลังฝึก [0.021 0.982 0.982 0.019]


In [54]:
def train_xor(lr, epochs=20000, seed=1):
    rng = np.random.default_rng(seed)
    X = np.array([[0,0],[0,1],[1,0],[1,1]], float)
    y = np.array([[0],[1],[1],[0]], float)
    W1 = rng.normal(0,1,(2,2)); b1 = np.zeros((1,2))
    W2 = rng.normal(0,1,(2,1)); b2 = np.zeros((1,1))
    sig = lambda z: 1/(1+np.exp(-z))
    for _ in range(epochs):
        h = sig(X@W1+b1); out = sig(h@W2+b2)
        d_out = (out-y)*out*(1-out)
        d_h = d_out@W2.T * h*(1-h)
        W2 -= lr*h.T@d_out; b2 -= lr*d_out.sum(0)
        W1 -= lr*X.T@d_h;  b1 -= lr*d_h.sum(0)
    h = sig(X@W1+b1); out = sig(h@W2+b2)
    return float(np.mean((out-y)**2)), np.round(out.ravel(), 3)

for lr in [0.01, 0.5, 5.0, 20.0]:
    L, preds = train_xor(lr)
    print(f"lr {lr:5.2f}  loss = {L:.4f}  คำตอบ = {preds}")

lr  0.01  loss = 0.2157  คำตอบ = [0.441 0.404 0.673 0.454]
lr  0.50  loss = 0.0001  คำตอบ = [0.013 0.989 0.989 0.011]
lr  5.00  loss = 0.0000  คำตอบ = [0.004 0.997 0.997 0.003]
lr 20.00  loss = 0.1443  คำตอบ = [0.    0.303 0.998 0.303]


สิ่งที่ต้องส่ง
ไฟล์ notebook หรือไฟล์ .py ที่รันครบด่านที่ 1 ถึง 7 พร้อมผลลัพธ์
คำตอบคำถามชวนคิดด่านที่ 2 และคำตอบด่านที่ 8 ทั้งสามข้อ

ด่าน 2
คำถามชวนคิดประจำด่าน
ก้าว 0.55 น่าสนใจที่สุด มันก้าวข้ามก้นหุบไปตกฝั่งตรงข้ามทุกก้าว (ลอง print ค่า x ทุกก้าวดู จะเห็นเครื่องหมายบวกลบสลับกัน) แต่สุดท้ายก็ถึงก้นหุบ ทำไมมันต่างจาก 1.05 ที่เด้งเหมือนกันแต่พัง (บอกใบ้ เทียบขนาดของ x แต่ละรอบว่าหดลงหรือโตขึ้น)

ผ่านด่านเมื่อ รันครบ 5 ค่า และตอบคำถามชวนคิดได้ด้วยหลักฐานจากการ print จริง


In [55]:
for lr in [0.01, 0.1, 0.45, 0.55, 1.05]:
    x = 8.0
    for _ in range(20):
        x = x - lr*2*x
    print(f"lr {lr:5.2f}  หลัง 20 ก้าว x = {x:.4f}")

lr  0.01  หลัง 20 ก้าว x = 5.3409
lr  0.10  หลัง 20 ก้าว x = 0.0922
lr  0.45  หลัง 20 ก้าว x = 0.0000
lr  0.55  หลัง 20 ก้าว x = 0.0000
lr  1.05  หลัง 20 ก้าว x = 53.8200


หลักฐาน

In [56]:
for lr in [0.01, 0.1, 0.45, 0.55, 1.05]:
    x = 8.0
    print(f"\nlr = {lr}")
    for step in range(20):
        x = x - lr * 2 * x
        print(f"ก้าว {step+1:2d}: x = {x:.6f}")


lr = 0.01
ก้าว  1: x = 7.840000
ก้าว  2: x = 7.683200
ก้าว  3: x = 7.529536
ก้าว  4: x = 7.378945
ก้าว  5: x = 7.231366
ก้าว  6: x = 7.086739
ก้าว  7: x = 6.945004
ก้าว  8: x = 6.806104
ก้าว  9: x = 6.669982
ก้าว 10: x = 6.536582
ก้าว 11: x = 6.405851
ก้าว 12: x = 6.277734
ก้าว 13: x = 6.152179
ก้าว 14: x = 6.029136
ก้าว 15: x = 5.908553
ก้าว 16: x = 5.790382
ก้าว 17: x = 5.674574
ก้าว 18: x = 5.561083
ก้าว 19: x = 5.449861
ก้าว 20: x = 5.340864

lr = 0.1
ก้าว  1: x = 6.400000
ก้าว  2: x = 5.120000
ก้าว  3: x = 4.096000
ก้าว  4: x = 3.276800
ก้าว  5: x = 2.621440
ก้าว  6: x = 2.097152
ก้าว  7: x = 1.677722
ก้าว  8: x = 1.342177
ก้าว  9: x = 1.073742
ก้าว 10: x = 0.858993
ก้าว 11: x = 0.687195
ก้าว 12: x = 0.549756
ก้าว 13: x = 0.439805
ก้าว 14: x = 0.351844
ก้าว 15: x = 0.281475
ก้าว 16: x = 0.225180
ก้าว 17: x = 0.180144
ก้าว 18: x = 0.144115
ก้าว 19: x = 0.115292
ก้าว 20: x = 0.092234

lr = 0.45
ก้าว  1: x = 0.800000
ก้าว  2: x = 0.080000
ก้าว  3: x = 0.008000
ก้าว  4: x = 0.000800


ตอบว่า
0.55 ถึงแม้ค่า x จะสลับบวก/ลบทุกก้าว แต่ขนาดของ x หดลงเรื่อย ๆ เพราะคูณด้วย -0.1 จึงเข้าใกล้ 0 และถึงก้นหุบได้ ส่วน 1.05 คูณด้วย -1.1 ทำให้ขนาดของ x ใหญ่ขึ้นทุกก้าว ถึงจะสลับเครื่องหมายเหมือนกัน แต่จะหนีออกจากก้นหุบ


```
# This is formatted as code
```



ด่านที่ 8คำถามปิดท้าย มุมยุค AI
ด่านนี้ไม่มีโค้ด ให้เขียนตอบสั้นๆ สามข้อด้วยภาษาตัวเอง


1.สามจังหวะของการฝึกเครือข่ายประสาทคืออะไร และในโค้ด PyTorch ของด่านที่ 6 แต่ละจังหวะอยู่บรรทัดไหน

2.โมเดลภาษาใหญ่ฝึกด้วยกลไกเดียวกับสมองจิ๋วของเรา จากสิ่งที่เห็นในแล็บนี้

3.อธิบายว่าทำไมโมเดลที่ฝึกแบบนี้ถึงตอบเก่งมาก แต่ก็ตอบผิดอย่างมั่นใจได้ด้วย (บอกใบ้ ตลอดการฝึก เราถามมันว่าอะไร และไม่เคยถามว่าอะไร)
ถ้าฝึกโมเดลแล้ว loss พุ่งขึ้นเรื่อยๆ แทนที่จะลด สาเหตุแรกที่ควรสงสัยคืออะไร และจะลองแก้ยังไง

ผ่านด่านเมื่อ ตอบครบสามข้อ โดยข้อ 2 ต้องอ้างถึง loss กับความจริง และข้อ 3 ต้องอ้างการทดลองของตัวเองจากด่านที่ 2 หรือ 7

ตอบ
1. สามจังหวะของการฝึกเครือข่ายประสาทคืออะไร

มี 3 จังหวะคือ คิดไปข้างหน้า → วัดความผิด → ปรับน้ำหนัก

ในโค้ดด่าน 6 คือ

คิดไปข้างหน้า → out = model(X)
วัดความผิด → loss = loss_fn(out, y)
ปรับน้ำหนัก → loss.backward() และ opt.step()

ตอบ

2. ทำไมโมเดลถึงตอบเก่ง แต่ตอบผิดอย่างมั่นใจได้

ตอนฝึก โมเดลจะพยายามลด loss โดยเทียบคำตอบที่ทำนายกับ ความจริง (y) แต่เราไม่เคยถามมันว่า “คำตอบนี้ถูกต้องตามความเป็นจริงหรือเปล่า” นอกเหนือจากข้อมูลที่ใช้ฝึก ดังนั้นถ้ามันเรียนรู้จากข้อมูลที่ผิดหรือไม่ครบ มันก็สามารถให้คำตอบที่ผิดแต่มั่นใจได้ เพราะมันเรียนรู้ให้ตรงกับรูปแบบในข้อมูล ไม่ได้เข้าใจความจริงเหมือนมนุษย์

ตอบ

3. ถ้า loss พุ่งขึ้นเรื่อย ๆ สาเหตุแรกที่ควรสงสัยคืออะไร

สิ่งแรกที่ควรสงสัยคือ learning rate หรือขนาดก้าวใหญ่เกินไป ทำให้การปรับน้ำหนักกระโดดข้ามจุดที่ดีไปเรื่อย ๆ

จากด่าน 7 เห็นว่า lr = 0.5 และ 5.0 ได้ loss ต่ำมาก แต่พอ lr = 20.0 loss กลับสูงขึ้นเป็น 0.1443 และคำตอบเพี้ยน แสดงว่าก้าวใหญ่เกินไปทำให้การฝึกพัง ดังนั้นถ้าเจอแบบนี้จะลอง ลด learning rate ลง ก่อนค่ะ